xarray/core/coordinates.py

```diff
@@ -796,6 +797,126 @@ def _ipython_key_completions_(self):
         ]
 
 
+class DataTreeCoordinates(DatasetCoordinates):
+    """
+    Dictionary like container for coordinates of a DataTree node (variables + indexes).
+    """
+
+    # TODO: This collection can be passed directly to the :py:class:`~xarray.Dataset`
+    # and :py:class:`~xarray.DataArray` constructors via their `coords` argument.
+    # This will add both the coordinates variables and their index.
+
+    # TODO: This only needs to be a separate class from `DatasetCoordinates` because DataTree nodes store their variables differently
+    # internally than how Datasets do, see https://github.com/pydata/xarray/issues/9203.
+
+    # TODO should inherited coordinates be here? It would be very hard to allow updating them...
+    # But actually maybe the ChainMap approach would make this work okay??
+
+    _data: DataTree
+
+    __slots__ = ("_data",)
+
+    def __init__(self, datatree: DataTree):
+        self._data = datatree
+
+    @property
+    def _names(self) -> set[Hashable]:
+        return set(self._data._coord_variables)
+
+    @property
+    def dims(self) -> Frozen[Hashable, int]:
+        # TODO is there a potential bug here? What happens if a dim is only present on data variables?
+        return self._data.dims
+
+    @property
+    def dtypes(self) -> Frozen[Hashable, np.dtype]:
+        """Mapping from coordinate names to dtypes.
+
+        Cannot be modified directly, but is updated when adding new variables.
+
+        See Also
+        --------
+        Dataset.dtypes
+        """
+        return Frozen({n: v.dtype for n, v in self._data._coord_variables.items()})
+
+    @property
+    def variables(self) -> Mapping[Hashable, Variable]:
+        variables = self._data._data_variables | self._data._coord_variables
+        return Frozen({k: v for k, v in variables.items() if k in self._names})
```

Comment:
I think this could just be `Frozen(self._data._coord_variables)` (we've already selected out all coordinate variables)

Reply:
done in https://github.com/pydata/xarray/pull/9451/commits/e8620cfc99c0949007d7f8f4e8c843e2834199d9